### PR TITLE
Implement Result Pattern

### LINKED ISSUE #18819: Result Pattern
As we know, currently we have `EmailResult`, `SmsResult`, both share the same construct for different services. Who knows, maybe in the future we will have `AbcResult`, `XyzResult` .. etc

The motivation is:

- DRY
- Create a type that represents a generic result
- Support generics, which is cool in case we want to return an object in both successful and failed operations

### PR DESCRIPTION
**Pull request by hishamco**
Fixes #18819

### CODE REVIEW

File: test/OrchardCore.Tests/Email/EmailTests.cs [hunk 1]
```diff
@@ -193,30 +193,6 @@ public async Task SendEmail_WithoutToAndCcAndBccHeaders_ShouldThrowsException()
         Assert.True(result.Errors.Any());
     }
 
-    [Fact]
```
Comment (Piedone) [inline test/OrchardCore.Tests/Email/EmailTests.cs:196]: Why delete this test?

File: src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    public IEnumerable<LocalizedString> Errors { get; protected set; }
+
+    public bool Succeeded { get; protected set; }
+
+    public static Result Success() =>  new()
+    {
+        Succeeded = true,
+    };
```
Comment (Piedone) [inline src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs]: This is creating an instance instead of the reused singleton in the other *Result classes. Why is this better?

File: src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+    protected Result(bool succeeded, IEnumerable<LocalizedString> errors)
+    {
+        Succeeded = succeeded;
+        Errors = errors;
+    }
+
+    public IEnumerable<LocalizedString> Errors { get; protected set; }
+
+    public bool Succeeded { get; protected set; }
```
Comment (Piedone) [inline src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs]: These can be `init` properties.

File: src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+    public static Result Success() =>  new()
+    {
+        Succeeded = true,
+    };
+
+    public static Result Failed(params LocalizedString[] errors) => new()
+    {
+        Succeeded = false,
```
Comment (Piedone) [inline src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs]: This is the default, no need to set `false`.

File: src/OrchardCore.Modules/OrchardCore.Email.Azure/Services/AzureEmailProviderBase.cs [hunk 5]
```diff
@@ -121,41 +122,41 @@ public virtual async Task<EmailResult> SendAsync(MailMessage message)
...
             _logger.LogError(ex, "An error occurred while sending an email using the Azure Email Provider.");
 
             // IMPORTANT: Do not expose ex.Message as it could contain the connection string in a raw format!
-            return EmailResult.FailedResult(string.Empty, S["An error occurred while sending an email."]);
+            return Result.Failed(S["An error occurred while sending an email."]);
         }
     }
 
-    private EmailMessage FromMailMessage(MailMessage message, Dictionary<string, IList<LocalizedString>> errors)
+    private EmailMessage FromMailMessage(MailMessage message, List<LocalizedString> errors)
```
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Email.Azure/Services/AzureEmailProviderBase.cs]: Without the dictionary, this potentially duplicates errors (though having the same e-mail address in multiple fields looks like an edge case).

File: src/OrchardCore.Modules/OrchardCore.Email.Azure/Services/AzureEmailProviderBase.cs [hunk 6]
```diff
@@ -121,41 +122,41 @@ public virtual async Task<EmailResult> SendAsync(MailMessage message)
...
         }
 
-        var errors = new Dictionary<string, IList<LocalizedString>>();
+        var errors = new List<LocalizedString>();
         var emailMessage = FromMailMessage(message, errors);
 
         if (errors.Count > 0)
         {
-            return EmailResult.FailedResult(errors);
+            return Result.Failed(errors.ToArray());
```
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Email.Azure/Services/AzureEmailProviderBase.cs]: You can have an overload in `Result` with `˙IEnumerable<LocalizedString> errors` instead, to prevent the needless array creation.

File: src/OrchardCore/OrchardCore.Email.Core/Services/EmailMessageValidator.cs [hunk 7]
```diff
@@ -21,50 +21,49 @@ public override Task ValidatingAsync(MailMessage message, MailMessageValidationC
             .Where(address => !_emailAddressValidator.Validate(address))
             .Select(address => S["Invalid email address for the sender: '{0}'.", address]);
 
-        AddError(context, nameof(message.From), invalidSender);
+        AddError(context, invalidSender);
```
Comment (Piedone) [inline src/OrchardCore/OrchardCore.Email.Core/Services/EmailMessageValidator.cs]: I wouldn't take removing the error key here so lightly. Did you look into why that was necessary in the first place?

File: src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs
Comment (Piedone) [inline src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs:1]: Add a note about this to the release notes.

File: test/OrchardCore.Tests/Email/EmailTests.cs (hunk 1 above)
Comment (hishamco) [reply test/OrchardCore.Tests/Email/EmailTests.cs:196]: Coz there's no response property anymore, you could use a generic type in case you want to return a success of failed object

File: src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs (hunk 2 above)
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs]: This need to capture the value from a `readonly` field instead. Let me check

File: src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs (hunk 3 above)
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs]: Yep, but while they are used by the subclasses, I used `get`

It's fine to add this

File: src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs:1]: I will add a note once there's no objection to the principle

File: src/OrchardCore/OrchardCore.Email.Core/Services/EmailMessageValidator.cs (hunk 7 above)
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Email.Core/Services/EmailMessageValidator.cs]: I will revisit this soon

File: src/OrchardCore.Modules/OrchardCore.Email.Azure/Services/AzureEmailProviderBase.cs (hunk 5 above)
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Email.Azure/Services/AzureEmailProviderBase.cs]: I'm not sure why this was introduced after the email refactoring, but I will have a look

File: src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs (hunk 4 above)
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs]: Copied from the old code, but I think you are right

File: test/OrchardCore.Tests/Email/EmailTests.cs (hunk 1 above)
Comment (Piedone) [reply test/OrchardCore.Tests/Email/EmailTests.cs:196]: That wasn't used at all, right?

File: src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs (hunk 3 above)
Comment (Piedone) [reply src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs]: `get` can remain, just `get; init;`.

File: test/OrchardCore.Tests/Email/EmailTests.cs (hunk 1 above)
Comment (hishamco) [reply test/OrchardCore.Tests/Email/EmailTests.cs:196]: Right, but I remembered there was a case to get the response from SMTP; I already handled this. Let me check

File: src/OrchardCore.Modules/OrchardCore.Email.Smtp/Services/SmtpEmailProviderBase.cs [hunk 8]
```diff
@@ -67,21 +68,21 @@ public virtual async Task<EmailResult> SendAsync(MailMessage message)
             {
                 var response = await SendOnlineMessageAsync(mimeMessage);
 
-                return EmailResult.GetSuccessResult(response);
+                return Result.Success(response);
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Email.Smtp/Services/SmtpEmailProviderBase.cs:75]: @Piedone, this is where the response is used

File: src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs (hunk 3 above)
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs]: I know what you mean :) But this might have a limitation coz the subclass will change this property only during the initialization

File: src/OrchardCore.Modules/OrchardCore.Email.Azure/Services/AzureEmailProviderBase.cs [hunk 9]
```diff
@@ -130,28 +135,42 @@ public virtual async Task<EmailResult> SendAsync(MailMessage message)
 
         if (errors.Count > 0)
         {
-            return EmailResult.FailedResult(errors);
+            var i = 0;
+            var resultErrors = new ResultError[errors.Count];
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Email.Azure/Services/AzureEmailProviderBase.cs]: Maybe linq would be simpler to read. Don't answer perf is a problem.

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Email.Azure/Services/AzureEmailProviderBase.cs]: Believe it or not, I used LINQ, but I think you will tell me PERF :)

File: src/OrchardCore.Modules/OrchardCore.Email.Smtp/Services/SmtpEmailProviderBase.cs (hunk 8 above)
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Email.Smtp/Services/SmtpEmailProviderBase.cs:75]: I'm not sure I understand what you mean.

File: src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs (hunk 3 above)
Comment (Piedone) [reply src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs]: I see. OK, let's leave them.

File: src/OrchardCore.Modules/OrchardCore.Email.Smtp/Services/SmtpEmailProviderBase.cs (hunk 8 above)
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Email.Smtp/Services/SmtpEmailProviderBase.cs:75]: You asked last time why I deleted the unit test, because there's no `Response` property in the current implementation, the operation can provide metadata in both cases. I'm working on some unit tests to cover the corner cases, to make everything clear

File: src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs [hunk 10]
```diff
@@ -2,36 +2,87 @@
...
         Succeeded = succeeded;
-        Errors = errors;
+        _errors = errors;
     }
 
-    public IEnumerable<LocalizedString> Errors { get; protected set; }
+    /// <summary>
+    /// Gets the collection of errors associated with the result.
+    /// </summary>
+    public IEnumerable<ResultError> Errors => _errors;
```
Comment (Piedone) [inline src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs:31]: Why use a manual backing field for this property?

File: src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs (hunk 4 above)
Comment (Piedone) [reply src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs]: Then please change it.

File: src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs [hunk 11]
```diff
@@ -2,36 +2,87 @@
...
+    /// <returns>A successful result instance with the specified value.</returns>
+    public static Result<TValue> Success<TValue>(TValue value)
+        => new Result<TValue>(value, true, []);
+
+    /// <summary>
+    /// Returns a failed result instance with the specified errors.
+    /// </summary>
+    /// <param name="errors">The errors that occurred during the operation.</param>
+    /// <returns>A failed result instance with the specified errors.</returns>
+    public static Result Failed(params ResultError[] errors)
```
Comment (Piedone) [inline src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs]: I don't think we need this, and the added complication of having to instantiate `ResultError`, for every single case with multiple errors. No need to have a key for each of them. My comment under https://github.com/OrchardCMS/OrchardCore/pull/18820#discussion_r2795080872 was just referring to that one use case. There you need a key, but only in the consumer.

Also see https://github.com/OrchardCMS/OrchardCore/pull/18820#discussion_r2795086029 what I still think need addressing.

File: src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs
Comment (Piedone) [reply src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs:1]: Still needed.

**Review by Piedone — CHANGES_REQUESTED** [review]

File: src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs:1]: Now it's time to add it

File: src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs (hunk 10 above)
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs:31]: Because it's read-only, the underlying store needs to be changed in `Failed(errors)`, so if it's `IList<ResultError>`, it will be a miss

File: src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs (hunk 11 above)
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs]: We could use `record` instead. Second, this might be useful, especially for `ModelState`, so no need to introduce a dictionary in each module that needs it

Something similar to [`IdentityError`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.identity.identityerror?view=aspnetcore-10.0)

Comment (hishamco) [reply src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs]: I revised the code; there's only one single result instance per multiple errors

Comment (Piedone) [reply src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs]: How would that be useful for `ModelState`? The one and only use case currently for a dictionary is for e-mails but that's very specific to that use case because there it makes sense to group validation errors per key. We don't need this complication otherwise. The less code the better.

File: src/docs/releases/3.0.0.md [hunk 12]
```diff
@@ -621,3 +621,40 @@ A new shape, `UserDisplayName` has been introduced to render a user's display na
 ### Content Manager Delete Behavior
 
 The `IContentManager.RemoveAsync()` method now supports cancelable deletes and returns a `bool` indicating whether the delete operation actually occurred or was cancelled by a handler.
+
+### Result Pattern
```
Comment (Piedone) [inline src/docs/releases/3.0.0.md:635]: I know that AI can generate this on a whim but it takes an effort to read it (if we want people to read the release notes, what we absolutely do), and I don't think we need a verbose explanation and usage tutorial for this, especially not in the release notes. So again, just a mention is enough, with pointers on where to see examples in the code (what won't be out of date if we change the code, unlike static documentation).

**Review by Piedone — CHANGES_REQUESTED** [review]

File: src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs (hunk 11 above)
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs]: This has been used for both Email and `RegisterUserTask`, I'm sure it could be used in SMS and other modules, especially if we want to add a key in `ModelState

As I said in my last comment, there's only one result per errors list

> The one and only use case currently for a dictionary is for e-mails, but that's very specific to that use case because there it makes sense to group validation errors per key. We don't need this complication otherwise. The less code, the better.

That's why Mike modified `EmailResult.Errors` to become `IReadOnlyDictionary<string, LocalizedString[]>`, look how it was https://github.com/OrchardCMS/OrchardCore/pull/12443/changes

IMHO, there's no complication here; otherwise, we will need to use `LocalizedString` and remove the key from the `ModelState` for consistency - which I don't prefer-

Comment (Piedone) [reply src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Result.cs]: OK, let's have it like this.

**Review by Piedone — APPROVED** [review]

**Comment by hishamco** [comment]
Do I need to write any extra unit tests, or are those enough?

**Comment by Piedone** [comment]
I think they're enough.

File: src/OrchardCore.Modules/OrchardCore.Email.Azure/Services/AzureEmailProviderBase.cs [hunk 13]
```diff
@@ -241,8 +252,6 @@ private EmailMessage FromMailMessage(MailMessage message, Dictionary<string, ILi
             {
                 errors.TryAdd(nameof(message.Attachments), []);
 
-                errors[nameof(message.Attachments)].Add(S["Unable to attach the file named '{0}' since its type is not supported.", attachment.Filename]);
```
Comment (gvkries) [inline src/OrchardCore.Modules/OrchardCore.Email.Azure/Services/AzureEmailProviderBase.cs:244]: Why remove this error message?

File: src/OrchardCore/OrchardCore.Email.Core/Services/DefaultEmailService.cs [hunk 14]
```diff
@@ -1,6 +1,9 @@
+using Microsoft.AspNetCore.DataProtection.KeyManagement;
```
Comment (gvkries) [inline src/OrchardCore/OrchardCore.Email.Core/Services/DefaultEmailService.cs:1]: Not needed.

File: src/OrchardCore/OrchardCore.Email.Core/Services/DefaultEmailService.cs [hunk 15]
```diff
@@ -1,6 +1,9 @@
...
 using Microsoft.Extensions.Localization;
 using Microsoft.Extensions.Logging;
+using OrchardCore.Infrastructure;
 using OrchardCore.Modules;
+using static System.Runtime.InteropServices.JavaScript.JSType;
```
Comment (gvkries) [inline src/OrchardCore/OrchardCore.Email.Core/Services/DefaultEmailService.cs:6]: Not needed.

Comment (hishamco) [reply src/OrchardCore/OrchardCore.Email.Core/Services/DefaultEmailService.cs:6]: Added accidentally by VS, nice catch!

Addressed in #18851
